src/modules/payload_deliverer/payload_deliverer.cpp

```diff
@@ -0,0 +1,255 @@
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+#include "payload_deliverer.h"
+
+PayloadDeliverer::PayloadDeliverer()
+	: ModuleParams(nullptr),
+	  ScheduledWorkItem(MODULE_NAME, px4::wq_configurations::lp_default)
+{
+}
+
+bool PayloadDeliverer::init()
+{
+	// Register callback that triggers the Run() function
+	if (!_vehicle_command_sub.registerCallback()) {
+		PX4_ERR("callback registration failed");
+		return false;
+	}
+
+	initialize_gripper();
+	return true;
+}
+
+bool PayloadDeliverer::initialize_gripper()
+{
+	// If gripper instance is invalid, try initializing it
+	if (!_gripper.is_valid() && _param_gripper_enable.get()) {
+		GripperConfig config{};
+		config.type = (GripperConfig::GripperType)_param_gripper_type.get();
+		config.sensor = GripperConfig::GripperSensorType::NONE; // Feedback sensor isn't supported for now
+		config.timeout_us = hrt_abstime(_param_gripper_timeout_s.get() * 1000000ULL);
+		_gripper.init(config);
+	}
+
+	// NOTE: Support for changing gripper sensor type / gripper type configuration when
+	// the parameter change is detected isn't added as we don't have actual use case for that
+	// yet!
+
+	if (!_gripper.is_valid()) {
+		PX4_DEBUG("Gripper object initialization invalid!");
+		return false;
+
+	} else {
+		_gripper.grab(); // Initialize to grab position
+		return true;
+	}
+}
+
+void PayloadDeliverer::parameter_update()
+{
+	// Call the ModuleParams function to update internal parameters
+	updateParams();
+	initialize_gripper();
+}
+
+void PayloadDeliverer::Run()
+{
+	if (should_exit()) {
+		ScheduleClear();
+		exit_and_cleanup();
+		return;
+	}
+
+	vehicle_command_s vcmd{};
+	hrt_abstime now = hrt_absolute_time();
+
+	if (_vehicle_command_sub.update(&vcmd)) {
+		update_gripper(now, &vcmd);
+
+	} else {
+		update_gripper(now);
+
+	}
+}
+
+void PayloadDeliverer::update_gripper(const hrt_abstime &now,  const vehicle_command_s *vehicle_command)
+{
+	if (!_param_gripper_enable.get()) {
+		PX4_WARN("Gripper isn't enabled, so the gripper functionality will not work!");
+		// If gripper isn't enabled, don't do anything
+		return;
+	}
```

Comment:
Why check for this, the module isn't expected to run if disabled.

Reply:
Good point. I instead changed to check if gripper instance is valid (since that actually can happen during runtime, if gripper setting isn't correct)